### ISSUE TITLE
can't get image rendered inline?

### ISSUE THREAD

**Issue by xmduhan** [gregsexton/ob-ipython#118]
When I try the sample code, It run. But the image does not render in inline, return array in result. How can I fix this?

`org file
  #+BEGIN_SRC ipython :session :exports both :results raw drawer
   %matplotlib inline
   import matplotlib.pyplot as plt
   import numpy as np
   plt.hist(np.random.randn(20000), bins=200)
   #+END_SRC

   #+RESULTS:
   :RESULTS:
   (array([   1.,    1.,    1.,    0.,    1.,    0.,    0.,    0.,    1.,
              1.,    1.,    1.,    0.,    2.,    1.,    0.,    2.,    1.,
              0.,    1.,    4.,    3.,    4.,    2.,    0.,    2.,    4.,
              0.,    3.,    7.,    9.,    7.,   12.,   13.,    9.,    8.,
             14.,   13.,   12.,   15.,   14.,   15.,   12.,   16.,   24.,
             20.,   25.,   19.,   34.,   24.,   42.,   38.,   52.,   39.,
             39.,   54.,   50.,   56.,   59.,   81.,   74.,   82.,   81.,
             95.,   81.,   77.,  105.,  111.,  124.,  103.,  128.,  134.,
            140.,  165.,  155.,  170.,  144.,  152.,  184.,  172.,  179.,
            208.,  219.,  211.,  238.,  234.,  230.,  237.,  261.,  266.,
            249.,  254.,  291.,  270.,  285.,  290.,  315.,  299.,  291.,
            289.,  291.,  295.,  309.,  299.,  330.,  268.,  315.,  327.,
            299.,  306.,  318.,  286.,  335.,  301.,  288.,  274.,  257.,
            239.,  259.,  284.,  251.,  234.,  230.,  260.,  234.,  214.,
            209.,  207.,  202.,  214.,  210.,  181.,  165.,  191.,  147.,
            130.,  177.,  130.,  134.,  124.,  132.,  131.,  101.,  103.,
            103.,   80.,   97.,   81.,   83.,   69.,   73.,   64.,   60.,
             55.,   37.,   62.,   45.,   36.,   41.,   32.,   35.,   29.,
             25.,   15.,   19.,   20.,    8.,   19.,    8.,   12.,    9.,
             11.,   14.,   11.,   11.,   11.,    5.,    7.,   13.,    4.,
              6.,   11.,    2.,    3.,    2.,    5.,    2.,    3.,    0.,
              2.,    0.,    1.,    0.,    2.,    1.,    0.,    1.,    0.,

... ... 

`

**Comment by gregsexton** [comment]
You edited to remove the `plt.show()`. With it, this works and produces only the figure for me. Without it, this returns the array (the result of the last line) and shows the figure afterwards. This is the new behaviour: we display the evaluted result of the last expression and display any figures generated.

Are you still having a problem? Without the `plt.show()` do you not get a figure at the end?

**Comment by xmduhan** [comment]
This is different between with plt.show() and and without show(), and I still not get a figure.
-------------------------------------------------------- with show() ---------------------------------------------------------------
   #+BEGIN_SRC ipython :session :exports both :results raw  
   %matplotlib inline
   import matplotlib.pyplot as plt
   import numpy as np
   plt.hist(np.random.randn(20000), bins=200)
   plt.show()
   #+END_SRC

   #+RESULTS:
   <matplotlib.figure.Figure at 0x7f999a975ed0>
   
------------------------------------------------- without show() ----------------------------------------------------------------------

   #+BEGIN_SRC ipython :session :exports both :results raw  
   %matplotlib inline
   import matplotlib.pyplot as plt
   import numpy as np
   plt.hist(np.random.randn(20000), bins=200)
   #+END_SRC

   #+RESULTS:
   (array([   1.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,
              0.,    0.,    0.,    0.,    0.,    0.,    1.,    0.,    0.,
              0.,    0.,    0.,    0.,    1.,    0.,    0.,    0.,    0.,

**Comment by gregsexton** [comment]
Maybe there's a problem writing the file? I think it tries to use the cwd. If you can give me something to go on here I'll try to repro and fix. As it is, I can't repro.

**Comment by ttoe** [comment]
I experience the exact same issue. Here some infos:

**OS:** macOS Sierra 10.12.6
**Emacs (emacs-plus from brew):** GNU Emacs 25.2.1 (x86_64-apple-darwin16.6.0, NS appkit-1504.83 Version 10.12.5 (Build 16F73)) of 2017-08-09
**IPython:** 6.1.0 (via pip3 install ipython)
**IPython has:** Python:  3.6.2
**ob-ipython:** 20170902.1253
**Org version:** 9.0.10 (release_9.0.10-5-g1654a5)
**ob-ipython setting:** (setq ob-ipython-command "/usr/local/bin/ipython")

Let me know, if something else might be useful to know.

**Comment by xmduhan** [comment]
@gregsexton
 Writing what file? 
Could you explan me a little about mechanism of ob-ipython Interaction with emacs, I will try get something useful for you to fix it.

**Comment by jamieforth** [comment]
Same issue here. Seems to be some kind of race condition in the Jupyter initialisation, the above example produces an image about half the time for me in a newly started Emacs process. I can't seem to reliably reproduce either behaviour.

Logs:
- Successful
```
executing Ipython code block...
Contacting host: localhost:9988
Type C-c C-c to view the image as text.
Saving file /tmp/resources/32513t6J.png...
Wrote /tmp/resources/32513t6J.png
Code block evaluation complete.
```

- Unsuccessful
```
executing Ipython code block...
Contacting host: localhost:9988
Code block evaluation complete.
```

Also providing a :file option is broken. The new renderer seems to write out to the provided path multiple times. Initially the png is rendered correctly, but then immediately overwritten by the values returned from the block (e.g. [test.txt](https://github.com/gregsexton/ob-ipython/files/1277483/test.txt)).

Debian testing
Emacs 25.2.2
Python 3.5.4rc1
IPython 6.1.0
ob-ipython 20170902.1253
Org mode version 9.0.10 (org-plus-contrib-20170904)
virtualenv 15.1.0
pyvenv 20170826.812
python-shell-prompt-detect-enabled: nil
python-shell-completion-native-disabled-interpreters: ("jupyter" "pypy" "ipython")

**Comment by concavegit** [comment]
Reverting to here made the images work again: https://github.com/gregsexton/ob-ipython/commit/920f28322e2064c60889535f42d0aedf7d5d6d7a

**Comment by zngguvnf** [comment]
> Reverting to here made the images work again: 920f283

Same here!

**Comment by gregsexton** [comment]
It's interesting that those that have listed specs are all on 6.1. I'm on 5.5. Don't suppose someone could quickly check this with ipython 5.5 and see if it works?

**Comment by dzop** [comment]
Hello, I was also experiencing the same problems using IPython 6.2.

I was able to fix it by prioritizing images in the list of display values sent to `ob-ipython--render`.

For me, the list looked like the following:

```emacs-lisp
'((text/plain . "<matplotlib.figure.Figure at 0x10c0e3160>")
  (image/png . "..."))
```
and it seems that the renderer stops processing the items in this list on the first one it successfully processes, so I was able to get inline images by sorting so that values with image types are first: dzop/ob-ipython@f817965

Hopefully this may fix the problem for everyone else.

P.S. Great package 👍
